### PR TITLE
Refactored Queue mechanism

### PR DESCRIPTION
**Pull request by JanPetr**
- Jobs are now split into 2 types - realtime and full reindex
- Realtime jobs have higher priority and will be processed before full reindex queue
- At least 1/3 of processed jobs will be jobs of full-reindex
- Removed raw SQL queries
- Added PHP doc annotations
- Slightly better readability

TODO:
- [x] Add "isFullReindex" flag to categories when #696 is merged

### CODE REVIEW

File: Model/Queue.php [hunk 1]
```diff
@@ -64,32 +82,35 @@ public function __construct(
...
         if ($this->configHelper->isQueueActive()) {
-            $this->insert($className, $method, $data, $data_size);
+            $this->db->insert($this->table, [
+                'created'   => date('Y-m-d H:i:s'),
+                'class'     => $className,
+                'method'    => $method,
+                'data'      => json_encode($data),
+                'data_size' => $dataSize,
+                'pid'       => null,
+                'is_full_reindex' => $isFullReindex ? 1 : 0,
```
Comment (damcou) [inline Model/Queue.php]: Is it a column that have already been added before ? I don't have it so I get an error on indexing

Comment (JanPetr) [reply Model/Queue.php]: /o\ Great catch, @damcou!! Thank you!
I added it manually to my DB and completely forgot to add it to `Upgrade` scripts of the extension /o\

Comment (JanPetr) [reply Model/Queue.php]: @damcou Fixed. You can go to `setup_module` table, lower version of Algolia module to `1.9.0` and run `php bin/magento setup:upgrade` command. It'll create the column :)

Comment (damcou) [reply Model/Queue.php]: Thanks :)

File: Model/Queue.php [hunk 2]
```diff
@@ -502,34 +585,58 @@ private function arrayMultisort()
...
+            $this->db->update($this->table, ['pid' => $pid], ['job_id IN (?)' => $jobsIds]);
         }
+    }
 
-        return $currentMax;
+    private function clearOldFailingJobs()
+    {
+        $retryLimit = $this->configHelper->getRetryLimit();
+
+        if ($retryLimit > 0 || true) {
```
Comment (damcou) [inline Model/Queue.php]: forgotten test ?

**Comment by JanPetr** [comment]
@damcou, I fixed your comments and improve jobs getting. Could you review? 🙏

**Comment by JanPetr** [comment]
> I just noticed that in the dedicated indexing queue page, the method column was empty for category jobs. I guess it's more related to the category indexing refactoring and the method name has changed?

Yes, @damcou. There have been some changes in categories which might not be 100% compatible with Indexing Queue page. I'll [check that](https://algolia.atlassian.net/browse/MAG-134) after further refactoring.

**Pull request merged**